keps/sig-auth/541-external-credential-providers/README.md

```diff
@@ -493,6 +494,114 @@ credentials).
 PEM format. The certificate must be valid at the time of execution. These
 credentials are used for mTLS handshakes.
 
+### Metrics
+
+As discussed [below](#rollout-upgrade-and-rollback-planning), there are 4
+primary metrics used by this feature set.
+
+```golang
+var (
+  execPluginCertTTL = k8smetrics.NewGaugeFunc(
+    k8smetrics.GaugeOpts{
+      Name: "rest_client_exec_plugin_ttl_seconds",
+      Help: "Gauge of the shortest TTL (time-to-live) of the client " +
+        "certificate(s) managed by the auth exec plugin. The value " +
+        "is in seconds until certificate expiry (negative if " +
+        "already expired). If auth exec plugins are unused or manage no " +
+        "TLS certificates, the value will be +INF.",
+    },
+    func() float64 {
+      if execPluginCertTTLAdapter.e == nil {
+        return math.Inf(1)
+      }
+      return execPluginCertTTLAdapter.e.Sub(time.Now()).Seconds()
+    },
+  )
+
+  execPluginCertRotation = k8smetrics.NewHistogram(
+    &k8smetrics.HistogramOpts{
+      Name: "rest_client_exec_plugin_certificate_rotation_age",
```

Comment:
I haven't read through this KEP in detail so apologies if I'm missing something obvious but I don't really understand the bucketing on this metric. As an operator, I don't really care when a cert was last rotated, but I *do* care how close it is to expiry, which can't be inferred from rotation time. A histogram distribution of when certs were last rotated on the cluster isn't particularly useful.

Reply:
OK, thanks for the feedback @ehashman. Appreciate you taking the time to look at this. This was another metric that has already shipped with this feature. I don't have a lot of context on why it was added originally, but can take a guess at the reason why it was added.

> care how close it is to expiry

Do you think that an operator can get this from the rest_client_exec_plugin_ttl_seconds metric? I.e., if the rest_client_exec_plugin_ttl_seconds is really low (i.e., single digit seconds), then the operator knows that the cert was really close to expiry. The information from rest_client_exec_plugin_certificate_rotation_age could then help the operator answer the question "is the reason why my cert is so close to expiry when rotated because the exec authenticator is rotating certs too slowly?"

I feel like I see 2 paths forward here: 1) update the bucketing for this metric (can we do that in a backwards-compatible way?), 2) add another metric that indicates to the operator how close certs are to expiry when they are rotated. I feel like rest_client_exec_plugin_ttl_seconds gets at 2, so I would lean towards 1 (updating the metric somehow in a backwards compatible way, if that is even possible), or simply leaving this metric alone.

Reply:
Another thing came to mind here - if the question we are trying to answer is "how close is my cert to expiry", should we bucket a metric by "percentage of lifetime at which a cert is rotated"? I don't think we could do this to this metric in a backwards compatible way, but maybe we could add another metric if need-be? If we use a percentage of the lifetime, then I think the metric will be more helpful to an operator that may have some certs that are longer lifetime (1 year) vs shorter (1 week).

Reply:
OK I realized now I got these metrics totally wrong. Slow moving this morning. Let me try to reply to your comment again.

> care how close it is to expiry

OK, so an operator wants to figure out how close the certificate was to expiry. I'm going to assume that the operator knows the TTL of their certs, e.g., 1 week (604800 seconds). The way the code is written right now, the cert will only be rotated 1) when we make a request to the API with this credential and we get a 401 back or 2) we are about to make a request with the creds and we see that the current time is past the expiration time of the creds (per the exec plugin). That says to me that the operator should expect that all cert rotations happen at around 604800 seconds, i.e., certs are _very_ close to expiry when they are rotated.

OK. I am going to stop typing now.

Reply:
> I do care how close it is to expiry

That's what the `rest_client_exec_plugin_ttl_seconds` metric is for. This metric is for detecting when rotations are happening much more frequently or less frequently than expected.

Reply:
Thanks both!